### ISSUE TITLE
Disable timeshift completely, as its apparently not working

### ISSUE THREAD

**Issue by 4f1sh3r** [rbuehlma/pvr.zattoo#134]
This issue is related to https://github.com/rbuehlma/pvr.zattoo/issues/99

I can reproduce this error on all of my RPis with an account directly from Zattoo (no reseller) on Kodi 18 and 19, so I assume that timeshift is simply not supported.

So I suggest to disable timeshift completely in the API, so that it is not visible to users.

Currently it looks like the plugin would support it and a normal user just would want to use it. This could lead into frustrations.

This has happened with my wife last week: she pressed "Pause" in the good hope that it will work. When she finished her work and came back to the TV, she pressed "Resume" and after 1min, it jumped to the current position. This was frustrating for her. if she had known that, she would have let it run alongside instead of not watching it at all.

Or do I have a misunderstanding and timeshift should actually work with a Zattoo premium account?
If so, I will happily provide debug logs.

Love your plugin btw. Its our single TV source since 2019. Thanks for the great work!

**Comment by rbuehlma** [comment]
I just tried this on my Zattoo Premium account and there, timeshift works without any issues.  As far as I know, the timeshift buffer contains around 1-3 hours. Therefore, if you pause the stream for a longer time, it might not work. In this case, you might  want to use replay.

**Comment by 4f1sh3r** [comment]
I can confirm it works if you just have a short timeframe, maybe 2-3 minutes. But I can definitively confirm that it's not working if you use a timeframe greater than 5 minutes.

Today I actually set up a fresh RPi4 with LibreElec 10.0.0.1 and nothing besides your pvr.zattoo plugin (and inputstream.adaptive) and I am able to directly reproduce the issue with my Zattoo Premium account.

Have you looked carefully during your test? Maybe you could test it with NTV or some other station which shows a clock the whole time, as the "jump" from the timeshift position to the current position happens so fast that its barely recognizable sometimes. You should also at least have 5mins in the timeshift buffer. It also doesn't jump to the live TV, but to maybe 1min behind the current time.

It actually behaves as follows:
- Turn on a station with a permanent clock visible
- Press pause, wait at least 5 or better 10min, then pless play
- you will then **see the correct, timeshifted program for maybe 1min**. 
- Then a sudden jump will happen **to approx 1min behind the current time**.
- Kodi will stay in this timeframe. You can see live TV shifted by approx. 1min in the past.
- Kodi will still show that its in timeshift mode and shows the 5 or 10min time shift, so no visible error

This is actually very strange behavior. I can make a video if its still unclear. :-)

**Comment by rbuehlma** [comment]
And it definitely works... for me ;) I paused for more than 13 minutes and was able to continue the stream for at least 4 minutes.
Still I believe you that its not working in your case. I suspect that Zattoo DE is behaving differently to Zattoo in Switzerland.

Now, to have this properly investigated, I will require a DE premium account for debugging and a DE IP.
Most probably, the outcome of this analysis, will be that the same behavior as already in place by 1&1 will apply which would then first require a change to the PVR api in Kodi prior fixing it in this addon.

**Comment by 4f1sh3r** [comment]
I just sent you an E-Mail with some informations, maybe I could help you out with a DE account (and probably IP as well) if that helps.

**Comment by rbuehlma** [comment]
@phunkyfish
i have analyze this issue:
when pausing a live stream, Zattoo now requires the client to send a POST request to a certain URL. When continuing the stream, the client needs to do another request to get a new URL for playing.

Currently, this is not possible for the addon since all the play/pause is done by the inputstream addon.

Some time ago, I have proposed an extension of the PVR API:

When ever some action (user or automated) happens on the stream initiated by a PVR addon (like play, pause, stop, end of stream, ...), a call is done to the addon. The addon should then have the possibility to execute some actions, either by returning the requested action to the caller, or through some async method (i would prefer the later but I'm not sure if that would fit the API design).

These actions could be:
- Play a stream providing an URL (including the properties)
- Stop playing
- For backwards compatibility: no action from the PVR addon to let inputstream do its work.

Such an extension would allow me to implement also other frequently requested features like continuing with the following epg entry after a replay show.

What is your opinion on such a feature?

**Comment by phunkyfish** [comment]
@rbuehlma yes, this is something we would like to do. @AlwinEsch and I have discussed how we would do user pause.

@AlwinEsch is currently working on a sandbox for binary addons. Part of this work would be an async mechanism so the PVR addon and inputstream instance could communicate. Right now this is not possible, but using something like grpc under the hood could allow for this. Hopefully allowing multiple inputstreams (like for Picture in Picture) and this messaging can happen in Kodi 20+.

@AlwinEsch how much of this do you think is feasible in kodi 20?

**Comment by AlwinEsch** [comment]
Not sure whether it will come in 20, there are always other things in between (e.g. font requests or multiple instances at PVR, requested by @ksooo). Hope to get my current requests in soon and bring the next ones regarding sandbox.
I actually want to finally get it ready to continue working on my web browser project (where sandbox would be great) :sweat_smile:.

**Comment by phunkyfish** [comment]
@rbuehlma so realistically it could be kodi 21. What we are really talking about here is an extension of the PVR API and the inputstream API.

How about we create an issue proposing the changes and how we would like it to work?

For instance clicking pause on the UI could potentially be overridden by the addon or have a default behaviour. Whereas the end of a stream is purely an event from an inputstream (or video player) without a default. Describing these things and having a roughly idea of how they would work would be useful.

**Comment by rbuehlma** [comment]
I have created a feature request over here:

https://forum.kodi.tv/showthread.php?tid=366162&pid=3077270

I hope this is the right place for it. github issues seem only to be for bug reports.
